### PR TITLE
Notifications rework

### PR DESCRIPTION
**Pull request by jeabakker**
ToDo
- [x] tests
- [x] muting notifications

### CODE REVIEW

File: engine/classes/Elgg/Traits/Entity/Subscriptions.php [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+	 *
+	 * @throws InvalidParameterException
+	 * @return string[]
+	 */
+	protected function normalizeSubscriptionMethods($methods = []): array {
+		if (!is_string($methods) && !is_array($methods)) {
+			$dbt = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2);
+			$caller = $dbt[1]['function'] ?? 'unknown';
+			
+			throw new InvalidParameterException(elgg_echo('Entity:Subscriptions:InvalidMethodsException', [$caller]));
```
Comment (jdalsem) [inline engine/classes/Elgg/Traits/Entity/Subscriptions.php:185]: change to invalidargumentexception (in whole file)

Comment (jeabakker) [reply engine/classes/Elgg/Traits/Entity/Subscriptions.php]: done

File: engine/upgrades.php [hunk 2]
```diff
@@ -10,4 +10,5 @@
 	\Elgg\Upgrades\AlterDatabaseToMultiByteCharset::class,
 	\Elgg\Upgrades\DeleteDiagnosticsPlugin::class,
 	\Elgg\Upgrades\NotificationsPrefix::class,
+	\Elgg\Upgrades\ChangeUserNotificationSettingsNamespace::class,
```
Comment (jdalsem) [inline engine/upgrades.php:13]: should be placed alphabetically i believe

File: languages/en.php [hunk 3]
```diff
@@ -1657,7 +1666,19 @@
...
 	'entity:delete:fail' => '%s could not be deleted.',
-
+	
+	'entity:subscribe:success' => "You've successfully subscribed to %s",
+	'entity:subscribe:fail' => "An error occured while subscribing to %s",
+	
+	'entity:unsubscribe:success' => "You've successfully unsubscribed from %s",
+	'entity:unsubscribe:fail' => "An error occured while unsubscribing from %s",
+	
+	'entity:mute:success' => "You've successfully muted notifications about %s",
```
Comment (jdalsem) [inline languages/en.php:1676]: notifications about should be notifications of (multiple lines)

File: mod/friends/classes/Elgg/Friends/Relationships.php [hunk 4]
```diff
@@ -17,7 +17,7 @@ class Relationships {
 	 * @return void
 	 * @since 3.2
 	 */
-	public static function createFriendRelationship(\Elgg\Event $event) {
```
Comment (jdalsem) [inline mod/friends/classes/Elgg/Friends/Relationships.php:20]: rename should be documented

File: mod/groups/classes/Elgg/Groups/Relationships.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+class Relationships {
+	
+	/**
+	 * Remove notification subscriptions when a user leaves a group
+	 *
+	 * @param \Elgg\Event $event 'delete', 'relationship'
+	 *
+	 * @return void
+	 */
+	public static function removeGroupNotificationSubscription(\Elgg\Event $event): void {
```
Comment (jdalsem) [inline mod/groups/classes/Elgg/Groups/Relationships.php:20]: function name should be plural subscriptions

File: docs/appendix/upgrade-notes/3.x-to-4.0.rst [hunk 6]
```diff
@@ -99,6 +99,19 @@ Multiple Recipients
...
+A generic storage for notification preferences has been introduced in ``\ElggUser::setNotificationSetting()`` and ``\ElggUser::getNotificationSettings()``, 
+the notification settings now have a 'purpose'. 
+
+For example ``group_join`` can be used to manage the default subscription you get with a group when you join the group.
+
+The Notifications plugin has generic handling of displaying and saving the settings.
+To display the setting extend the view ``notifications/settings/records`` (plural) with a view which uses ``notifications/settings/record`` (singular).
+
+When requesting notification settings other than the default setting, if the user hasn't saved a setting yet it'll fall back to the default notification settings.
+
```
Comment (jdalsem) [inline docs/appendix/upgrade-notes/3.x-to-4.0.rst:114]: No mention about removal of friends collections notification management

File: mod/notifications/elgg-plugin.php [hunk 7]
```diff
@@ -13,13 +15,19 @@
 		'settings:notification:personal' => [
 			'path' => '/notifications/personal/{username?}',
 			'resource' => 'notifications/index',
+			'middelware' => [
```
Comment (jdalsem) [inline mod/notifications/elgg-plugin.php:18]: middleware

File: mod/notifications/elgg-plugin.php [hunk 8]
```diff
@@ -13,13 +15,19 @@
...
+				UserPageOwnerCanEditGatekeeper::class,
+			],
 		],
 		'settings:notification:groups' => [
 			'path' => '/notifications/group/{username?}',
 			'resource' => 'notifications/groups',
 			'required_plugins' => [
 				'groups',
 			],
+			'middelware' => [
```
Comment (jdalsem) [inline mod/notifications/elgg-plugin.php:28]: middleware

File: mod/notifications/languages/en.php [hunk 9]
```diff
@@ -7,18 +7,16 @@
...
 	'notifications:subscriptions:changesettings' => 'Notifications',
-	'notifications:subscriptions:changesettings:groups' => 'Group notifications',
-
-	'notifications:subscriptions:title' => 'Notifications per user',
-	'notifications:subscriptions:description' => 'To receive notifications from your friends (on an individual basis) when they create new content, find them below and select the notification method you would like to use.',
+	
+	'notifications:subscriptions:users:title' => 'Notifications per user',
+	'notifications:subscriptions:users:description' => 'To receive notifications from your friends (on an individual basis) when they create new content, find them below and select the notification method you would like to use.',
 
+	'notifications:subscriptions:groups:title' => 'Group notifications',
```
Comment (jdalsem) [inline mod/notifications/languages/en.php:19]: should this not come from groups?

File: mod/notifications/views/default/forms/notifications/subscriptions/groups.php [hunk 10]
```diff
@@ -4,6 +4,10 @@
  *
  * @uses $vars['user'] ElggUser
  */
+
```
Comment (jdalsem) [inline mod/notifications/views/default/forms/notifications/subscriptions/groups.php:7]: move to groups???

File: mod/site_notifications/elgg-plugin.php [hunk 11]
```diff
@@ -72,4 +72,11 @@
 			],
 		],
 	],
+	'events' => [
+		'create' => [
+			'user' => [
+				'Elgg\SiteNotifications\Users::enableSiteNotifications' => [],
```
Comment (jdalsem) [inline mod/site_notifications/elgg-plugin.php:78]: maybe register low priority?

**Review by jdalsem — CHANGES_REQUESTED** [review]

File: engine/classes/Elgg/Comments/AutoSubscribeHandler.php [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+	 */
+	public function __invoke(\Elgg\Event $event): void {
+		$object = $event->getObject();
+		if (!$object instanceof \ElggComment) {
+			return;
+		}
+		
+		$owner = $object->getOwnerEntity();
+		$container = $object->getContainerEntity();
+		if (!$owner instanceof \ElggUser || !$container instanceof \ElggEntity) {
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply engine/classes/Elgg/Comments/AutoSubscribeHandler.php:29]: nope

File: mod/friends/classes/Elgg/Friends/Relationships.php (hunk 4 above)
Comment (jeabakker) [reply mod/friends/classes/Elgg/Friends/Relationships.php:20]: was documented

File: mod/notifications/views/default/forms/notifications/subscriptions/groups.php (hunk 10 above)
Comment (jeabakker) [reply mod/notifications/views/default/forms/notifications/subscriptions/groups.php:7]: yes not in the PR

**Review by jdalsem — APPROVED** [review]

**Pull request merged**